File: bin/data/AddRotationBehaviour.lua
--
-- Created by IntelliJ IDEA.
-- User: damian
-- Date: 01/01/17
-- Time: 13:50
-- To change this template use File | Settings | File Templates.
--

require "Math"
require "Vector3"
require "Quaternion"
require "GLMUtils"

AddRotationBehaviour = {}

function AddRotationBehaviour:new(target, angularVelocity)
    o = {target=target, angularVelocity=angularVelocity}
    setmetatable(o, self)
    self.__index = self
    return o
end

function AddRotationBehaviour:Update(dt)

    local deltaRotation = kll.gquat(self.angularVelocity * dt)

    local newRotation = self.target:GetOrientation() * deltaRotation

    self.target:SetOrientation(newRotation)

end

function AddRotationBehaviour:ShouldObjectBeDestroyed()
    return false
end

function AddRotationBehaviour:GetTarget()
    return self.target
end

File: src/kll/Behaviours/AddRotationBehaviour.cpp
//
// Created by Damian Stewart on 11/12/2016.
//

#include <glm/gtc/quaternion.hpp>
#include "AddRotationBehaviour.h"
#include "../Objects/Object.h"

kll::AddRotationBehaviour::AddRotationBehaviour(Object *target, vec3 angularVelocity)
        : mAngularVelocity(angularVelocity)
{
    mTarget = target;

}

void kll::AddRotationBehaviour::Update(float dt)
{
    auto deltaRotation = glm::quat(mAngularVelocity * dt);

    auto newRotation = mTarget->GetOrientation() * deltaRotation;
    mTarget->SetOrientation(newRotation);
}
